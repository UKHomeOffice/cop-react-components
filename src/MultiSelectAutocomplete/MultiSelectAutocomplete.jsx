/**
 * EXPERIMENTAL
 * 
 * If this replaces the current "Autocomplete" in the future,
 * the "MultiSelectAutocomplete" should remain an alias or else
 * it would have to be a major release and not backwards compatible.
 */

import React, { useRef } from 'react';
import Select from 'react-select';
import PropTypes from 'prop-types';

import Readonly from '../Readonly';

import { getTemplates } from './MultiSelectAutocomplete.utils';

import { concatClasses } from '../utils/Utils';

import './MultiSelectAutocomplete.scss';

export const DEFAULT_CLASS = 'hods-multi-select-autocomplete';

const DEFAULT_VALUE = { value: "", label: "Select..." };
const GOVUK_COLOR_BLUE = '#1d70b8';
const GOV_COLOR_BLACK = '#000000';
const GOV_COLOR_WHITE = '#ffffff';
const GOV_COLOR_RED = '#d4351c';
const GOV_COLOR_NONE = '';

const MultiSelectAutocomplete = ({
  id,
  fieldId,
  disabled,
  error,
  readonly,
  item,
  value,
  options,
  isMulti,
  templates: _templates,
  onChange,
  onConfirm: _onConfirm,
  className: _className,
  ...attrs
}) => {
  const aacRef = useRef(null);
  const className = concatClasses(_className, error ? 'error' : undefined);
  const templates = getTemplates(_templates, item);

  const onItemSelected = (selection) => {
    if (typeof onChange === 'function') {
      onChange({
        target: {
          name: fieldId,
          value: selection !== DEFAULT_VALUE ? selection : undefined 
        }
      });
    }
  };

  const filterOptions = ({ label }, input) => {
    const lcLabel = label.toLowerCase();
    const lcQuery = input ? input.toLowerCase() : '';
    return lcLabel.includes(lcQuery)
  };

  /**
   * As this underlying component has two ways is handling is selected state, this method handles 
   * isSelected for single select.
   */
  const isSelected = (state) => {
    return state.data === state.selectProps.value;
  };

  /**
   * Provides another layer of customization for the autocomplete component (enables and disabled features, 
   * inclusing overriding some default behaviour).
   */
  const customComponents = { 
    DropdownIndicator:() => null,
    ClearIndicator:() => null,
    IndicatorSeparator:() => null,
    NoOptionsMessage:() => 'No results found',
  };

  /**
   * Primarily styles the options items within the menu list based on the state of the option
   * (Some of the desired behaviour can be achieve through regular css styling howver,
   * this allows for handling multiple conditions).
   */
  const customStyles = {
    option: (provided, state) => ({
      ...provided,
      ...(((state.isSelected || isSelected(state)) || (!state.isFocused && !state.isSelected)) 
        && {
        backgroundColor: GOV_COLOR_NONE,
        color: GOV_COLOR_BLACK
      }),
      ...(((state.isFocused && (state.isSelected || isSelected(state))) 
      || (state.isFocused && (!state.isSelected || !isSelected(state)))) && {
        backgroundColor: GOVUK_COLOR_BLUE,
        color: GOV_COLOR_WHITE
      }),
    }),
  };

  /**
   * Applies styling to the menu list but primarily sets the highlight colour which
   * appear on the menu list (for when travesing the available options and selecting an option),
   * the remove option button when in multi-select mode.
   */
  const customTheme = (theme) => ({
    ...theme,
    borderRadius: 0,
    colors: {
      ...theme.colors,
      primary50: GOVUK_COLOR_BLUE,
      primary25: GOVUK_COLOR_BLUE,
      dangerLight: GOV_COLOR_RED
    },
  });

  if (readonly) {
    let displayValue = '';
    if(Array.isArray(value)){
      const itemLabels = value.map((v) => templates.inputValue(v)); 
      const lastItem = itemLabels.pop();
      if(itemLabels.length > 0) {
        displayValue = `${itemLabels.join(', ')} and ${lastItem}`;
      } else {
        displayValue = lastItem;
      }
    } else {
      displayValue = templates.inputValue(value);
    }
    return (
      <Readonly id={id} className={className} {...attrs}>
        {displayValue}
      </Readonly>
    );
  }

  return (
    <div className={`${DEFAULT_CLASS}__outer-wrapper ${className ?? ''}`}>
      <Select
      components={customComponents}
      styles={customStyles}
      theme={customTheme}
      ref={aacRef}
      id={id}
      fieldId={fieldId}
      classNamePrefix={className}
      value={value}
      filterOption={filterOptions}
      onChange={onItemSelected}
      {...attrs}
      options={[
        isMulti ? undefined : DEFAULT_VALUE,
        ...options
      ].filter(o => !!o)}
      isMulti={isMulti}
      openMenuOnClick={false}
      menuShouldScrollIntoView={false}
      isDisabled={disabled}
      isClearable={true}
      />
    </div>
  )
};

MultiSelectAutocomplete.propTypes = {
  id: PropTypes.string.isRequired,
  fieldId: PropTypes.string.isRequired,
  disabled: PropTypes.bool,
  error: PropTypes.string,
  readonly: PropTypes.bool,
  /** The structure of the item. If this is not provided, it is assumed to be a string. */
  item: PropTypes.shape({
    value: PropTypes.string.isRequired,
    label: PropTypes.string.isRequired,
    format: PropTypes.string
  }),
  /** The selected item. */
  value: PropTypes.any,
  /** The autocomplete options */
  options: PropTypes.arrayOf(PropTypes.object),
  isMulti: PropTypes.bool,
  /** Functions for formatting the labels that appear in the options menu and in the input when a selection is made. */
  templates: PropTypes.shape({ inputValue: PropTypes.func.isRequired, suggestion: PropTypes.func.isRequired }),
  /** Handler for when the value changes. */
  onChange: PropTypes.func,
  /** Deprecated. Use onChange. */
  onConfirm: PropTypes.func,
  className: PropTypes.string
};

MultiSelectAutocomplete.defaultProps = {
  classBlock: DEFAULT_CLASS,
  value: DEFAULT_VALUE,
};

export default MultiSelectAutocomplete;
