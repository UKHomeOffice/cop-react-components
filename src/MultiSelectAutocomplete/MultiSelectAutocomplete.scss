$govuk-font-family: 'Roboto', arial, sans-serif;

@import "node_modules/govuk-frontend/govuk/_base";
@import "node_modules/govuk-frontend/govuk/components/input";
@import "node_modules/govuk-frontend/govuk/components/hint";

.hods-multi-select-autocomplete__outer-wrapper {
  
  .hods-multi-select-autocomplete {
  
    .error__control,
    &__control {
      width: 100%;
      border: 2px solid govuk-colour("black");
      border-radius: 0;
      padding: 0;
      font-size: 19px;
      box-sizing: border-box;
      transition: none;

      &--is-focused {
        outline: 3px solid govuk-colour("yellow") !important;
        outline-offset: 0;
        box-shadow: inset 0 0 0 2px;
      }

      &--is-disabled {
        background-color: #efefef4d;
      }

      &:hover {
        border: 2px solid $govuk-input-border-colour;
      }
    }
    
    .error__value-container,
    &__value-container {
      padding: 0.3125rem;
      font-size: 1.1875rem;
      line-height: 1.31579rem;
      
      &--is-multi {
        display: flex;
        flex-wrap: wrap;
      }
    }
  
    &.error__menu,
    &__menu {
      background-color: govuk-colour("white");
      border: 2px solid $govuk-input-border-colour;
      border-top: 0;
      color: govuk-colour("black");
      margin: 0;
      max-height: 324px;
      overflow-x: hidden;
      padding: 0;
      list-style-type: none;
      display: block;
      position: relative;
    }

    &.error__menu-list,
    &__menu-list {
      padding-bottom: 0;
      padding-top: 0;
    }
  
    &.error__single-value,
    &__single-value {
      color: govuk-colour("black");
      margin-left: 0;
      margin-right: 0;
      position: relative;
      display: inline-block;
      top: 12px;
    }

    &.error__multi-value__label,
    &__multi-value__label {
      padding-right: 6px;
    }

    &.error__multi-value__remove,
    &__multi-value__remove {
      &:hover {
        color: govuk-colour("white");
        background-color: govuk-colour("red");
      }
    }

    &__input {
      padding-left: 0px;
    }

    &.error__control {
      border: $govuk-border-width-form-element-error solid $govuk-error-colour;
      
      &--is-focused {
        border-color: $govuk-input-border-colour !important;
        outline: 3px solid govuk-colour("yellow") !important;
        outline-offset: 0;
        box-shadow: inset 0 0 0 2px;
      }
  
      &:hover:not(.error__control--is-focused) {
        border: 2px solid $govuk-error-colour;
      }
    }

    &.error__menu {
      position: inherit !important;
    }
  }
}
