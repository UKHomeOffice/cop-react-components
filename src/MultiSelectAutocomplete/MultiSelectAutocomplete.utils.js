import { interpolateString } from '../utils/Utils';

const getItemLabel = (property) => {
  return (item) => {
    return item && item[property] ? item[property] : '';
  }
};

const interpolateLabel = (format) => {
  return (item) => {
    return interpolateString(format, item).trim();
  }
};

const itemLabel = (structure) => {
  if (structure) {
    return structure.format ? interpolateLabel(structure.format) : getItemLabel(structure.label);
  }
  return (item) => item || '';
};

const isFunction = (fn) => {
  return typeof fn === 'function';
};

export const getTemplates = (templates, itemStructure) => {
  if (templates) {
    // If we only have one property, mirror it on the other one.
    if (isFunction(templates.inputValue)) {
      if (!isFunction(templates.suggestion)) {
        templates.suggestion = templates.inputValue;
      }
      return templates;
    } else if (isFunction(templates.suggestion)) {
      templates.inputValue = templates.suggestion;
      return templates;
    }
  }
  // We don't have a usable templates object so build a whole new one.
  const label = itemLabel(itemStructure);
  return {
    inputValue: label,
    suggestion: label
  };
};
