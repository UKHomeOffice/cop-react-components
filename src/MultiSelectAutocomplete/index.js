import MultiSelectAutocomplete from "./MultiSelectAutocomplete";

export default MultiSelectAutocomplete;
