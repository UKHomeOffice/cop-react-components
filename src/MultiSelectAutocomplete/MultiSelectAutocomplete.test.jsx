import React from 'react';
import MultiSelectAutocomplete, { DEFAULT_CLASS } from './MultiSelectAutocomplete';
import renderer from 'react-test-renderer';

describe('MultiSelectAutocomplete', () => {
  const AUTOCOMPLETE_ID = 'autocompleteId';
  const AUTOCOMPLETE_FIELD_ID = 'autocompleteFieldId';
  const AUTOCOMPLETE_TEST_ID = 'autocompleteTestId';
  const VALUES = [
    { value: "alpha", label: "alpha" },
    { value: "bravo", label: "bravo" }
  ]

  it('should render when disabled flag is set', async () => {
    const OPTIONS = {
      id: AUTOCOMPLETE_ID,
      'data-testid': AUTOCOMPLETE_TEST_ID,
      fieldId: AUTOCOMPLETE_FIELD_ID,
      value: VALUES[0],
      options: VALUES,
      disabled: true
    };

    const tree = renderer.create(<MultiSelectAutocomplete {...OPTIONS} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render a fully wired-up autocomplete component', async () => {
    const OPTIONS = {
      id: AUTOCOMPLETE_ID,
      fieldId: AUTOCOMPLETE_FIELD_ID,
      value: [],
      options: VALUES,
      disabled: false,
      isMulti: false,
      className:DEFAULT_CLASS
    };
    
    const tree = renderer.create(<MultiSelectAutocomplete {...OPTIONS} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render an input with disabled flag set and the error flag set', async () => {
    const OPTIONS = {
      id: AUTOCOMPLETE_ID,
      fieldId: AUTOCOMPLETE_FIELD_ID,
      value: [],
      options: VALUES,
      disabled: true,
      isMulti: false,
      className:DEFAULT_CLASS,
      error: 'Error'
    };

    const tree = renderer.create(<MultiSelectAutocomplete {...OPTIONS} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('should render an autocomplete component with an error control', async () => {
    const OPTIONS = {
      id: AUTOCOMPLETE_ID,
      'data-testid': AUTOCOMPLETE_TEST_ID,
      fieldId: AUTOCOMPLETE_FIELD_ID,
      value: VALUES[0],
      options: VALUES,
      disabled: false,
      error: 'Error'
    };

    const tree = renderer.create(<MultiSelectAutocomplete {...OPTIONS} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`should render an autocomplete component with a preselected value of "${VALUES[0].value}"`, async () => {
    const OPTIONS = {
      id: AUTOCOMPLETE_ID,
      fieldId: AUTOCOMPLETE_FIELD_ID,
      value: [VALUES[0]],
      options: VALUES,
      disabled: false,
      isMulti: false,
      className:DEFAULT_CLASS
    };
    const tree = renderer.create(<MultiSelectAutocomplete {...OPTIONS} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`should render an autocomplete component with a preselected value of "${VALUES[1].value}"`, async () => {
    const OPTIONS = {
      id: AUTOCOMPLETE_ID,
      fieldId: AUTOCOMPLETE_FIELD_ID,
      value: [VALUES[1]],
      options: VALUES,
      disabled: false,
      isMulti: false,
      className:DEFAULT_CLASS
    };
    const tree = renderer.create(<MultiSelectAutocomplete {...OPTIONS} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`should render an autocomplete component in multi select mode with preselected value of ${VALUES[0].value}`, async () => {
    
    const OPTIONS = {
      id: AUTOCOMPLETE_ID,
      fieldId: AUTOCOMPLETE_FIELD_ID,
      value: [VALUES[0]],
      options: VALUES,
      disabled: false,
      isMulti: true,
      className:DEFAULT_CLASS,
      onChange: jest.fn()
    };
    
    const tree = renderer.create(<MultiSelectAutocomplete {...OPTIONS} />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it(`should render an autocomplete component in multi select mode with preselected values of ${VALUES[0].value}, ${VALUES[1].value} `, async () => {
    
    const OPTIONS = {
      id: AUTOCOMPLETE_ID,
      fieldId: AUTOCOMPLETE_FIELD_ID,
      value: [...VALUES],
      options: VALUES,
      disabled: false,
      isMulti: true,
      className:DEFAULT_CLASS,
      onChange: jest.fn()
    };
    
    const tree = renderer.create(<MultiSelectAutocomplete {...OPTIONS} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
