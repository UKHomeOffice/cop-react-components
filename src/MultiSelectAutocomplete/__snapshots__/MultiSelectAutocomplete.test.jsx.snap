// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MultiSelectAutocomplete should render a fully wired-up autocomplete component 1`] = `
<div
  className="hods-multi-select-autocomplete__outer-wrapper hods-multi-select-autocomplete"
>
  <div
    className=" css-b62m3t-container"
    id="autocompleteId"
    onKeyDown={[Function]}
  >
    <span
      className="css-1f43avz-a11yText-A11yText"
      id="react-select-3-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      className="css-1f43avz-a11yText-A11yText"
    />
    <div
      className="hods-multi-select-autocomplete__control css-x19aeh-control"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className="hods-multi-select-autocomplete__value-container css-319lph-ValueContainer"
      >
        <div
          className="hods-multi-select-autocomplete__placeholder css-14el2xx-placeholder"
          id="react-select-3-placeholder"
        >
          Select...
        </div>
        <div
          className="hods-multi-select-autocomplete__input-container css-6j8wv5-Input"
          data-value=""
        >
          <input
            aria-autocomplete="list"
            aria-describedby="react-select-3-placeholder"
            aria-expanded={false}
            aria-haspopup={true}
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            className="hods-multi-select-autocomplete__input"
            disabled={false}
            id="react-select-3-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            role="combobox"
            spellCheck="false"
            style={
              Object {
                "background": 0,
                "border": 0,
                "color": "inherit",
                "font": "inherit",
                "gridArea": "1 / 2",
                "label": "input",
                "margin": 0,
                "minWidth": "2px",
                "opacity": 1,
                "outline": 0,
                "padding": 0,
                "width": "100%",
              }
            }
            tabIndex={0}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="hods-multi-select-autocomplete__indicators css-1hb7zxy-IndicatorsContainer"
      />
    </div>
  </div>
</div>
`;

exports[`MultiSelectAutocomplete should render an autocomplete component in multi select mode with preselected value of alpha 1`] = `
<div
  className="hods-multi-select-autocomplete__outer-wrapper hods-multi-select-autocomplete"
>
  <div
    className=" css-b62m3t-container"
    id="autocompleteId"
    onKeyDown={[Function]}
  >
    <span
      className="css-1f43avz-a11yText-A11yText"
      id="react-select-8-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      className="css-1f43avz-a11yText-A11yText"
    />
    <div
      className="hods-multi-select-autocomplete__control css-x19aeh-control"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className="hods-multi-select-autocomplete__value-container hods-multi-select-autocomplete__value-container--has-value css-319lph-ValueContainer"
      >
        <div
          className="hods-multi-select-autocomplete__single-value css-qc6sy-singleValue"
        >
          alpha
        </div>
        <div
          className="hods-multi-select-autocomplete__input-container css-6j8wv5-Input"
          data-value=""
        >
          <input
            aria-autocomplete="list"
            aria-expanded={false}
            aria-haspopup={true}
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            className="hods-multi-select-autocomplete__input"
            disabled={false}
            id="react-select-8-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            role="combobox"
            spellCheck="false"
            style={
              Object {
                "background": 0,
                "border": 0,
                "color": "inherit",
                "font": "inherit",
                "gridArea": "1 / 2",
                "label": "input",
                "margin": 0,
                "minWidth": "2px",
                "opacity": 1,
                "outline": 0,
                "padding": 0,
                "width": "100%",
              }
            }
            tabIndex={0}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="hods-multi-select-autocomplete__indicators css-1hb7zxy-IndicatorsContainer"
      />
    </div>
  </div>
</div>
`;

exports[`MultiSelectAutocomplete should render an autocomplete component in multi select mode with preselected values of alpha, bravo  1`] = `
<div
  className="hods-multi-select-autocomplete__outer-wrapper hods-multi-select-autocomplete"
>
  <div
    className=" css-b62m3t-container"
    id="autocompleteId"
    onKeyDown={[Function]}
  >
    <span
      className="css-1f43avz-a11yText-A11yText"
      id="react-select-9-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      className="css-1f43avz-a11yText-A11yText"
    />
    <div
      className="hods-multi-select-autocomplete__control css-x19aeh-control"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className="hods-multi-select-autocomplete__value-container hods-multi-select-autocomplete__value-container--has-value css-319lph-ValueContainer"
      >
        <div
          className="hods-multi-select-autocomplete__single-value css-qc6sy-singleValue"
        >
          alpha
        </div>
        <div
          className="hods-multi-select-autocomplete__input-container css-6j8wv5-Input"
          data-value=""
        >
          <input
            aria-autocomplete="list"
            aria-expanded={false}
            aria-haspopup={true}
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            className="hods-multi-select-autocomplete__input"
            disabled={false}
            id="react-select-9-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            role="combobox"
            spellCheck="false"
            style={
              Object {
                "background": 0,
                "border": 0,
                "color": "inherit",
                "font": "inherit",
                "gridArea": "1 / 2",
                "label": "input",
                "margin": 0,
                "minWidth": "2px",
                "opacity": 1,
                "outline": 0,
                "padding": 0,
                "width": "100%",
              }
            }
            tabIndex={0}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="hods-multi-select-autocomplete__indicators css-1hb7zxy-IndicatorsContainer"
      />
    </div>
  </div>
</div>
`;

exports[`MultiSelectAutocomplete should render an autocomplete component with a preselected value of "alpha" 1`] = `
<div
  className="hods-multi-select-autocomplete__outer-wrapper hods-multi-select-autocomplete"
>
  <div
    className=" css-b62m3t-container"
    id="autocompleteId"
    onKeyDown={[Function]}
  >
    <span
      className="css-1f43avz-a11yText-A11yText"
      id="react-select-6-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      className="css-1f43avz-a11yText-A11yText"
    />
    <div
      className="hods-multi-select-autocomplete__control css-x19aeh-control"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className="hods-multi-select-autocomplete__value-container hods-multi-select-autocomplete__value-container--has-value css-319lph-ValueContainer"
      >
        <div
          className="hods-multi-select-autocomplete__single-value css-qc6sy-singleValue"
        >
          alpha
        </div>
        <div
          className="hods-multi-select-autocomplete__input-container css-6j8wv5-Input"
          data-value=""
        >
          <input
            aria-autocomplete="list"
            aria-expanded={false}
            aria-haspopup={true}
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            className="hods-multi-select-autocomplete__input"
            disabled={false}
            id="react-select-6-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            role="combobox"
            spellCheck="false"
            style={
              Object {
                "background": 0,
                "border": 0,
                "color": "inherit",
                "font": "inherit",
                "gridArea": "1 / 2",
                "label": "input",
                "margin": 0,
                "minWidth": "2px",
                "opacity": 1,
                "outline": 0,
                "padding": 0,
                "width": "100%",
              }
            }
            tabIndex={0}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="hods-multi-select-autocomplete__indicators css-1hb7zxy-IndicatorsContainer"
      />
    </div>
  </div>
</div>
`;

exports[`MultiSelectAutocomplete should render an autocomplete component with a preselected value of "bravo" 1`] = `
<div
  className="hods-multi-select-autocomplete__outer-wrapper hods-multi-select-autocomplete"
>
  <div
    className=" css-b62m3t-container"
    id="autocompleteId"
    onKeyDown={[Function]}
  >
    <span
      className="css-1f43avz-a11yText-A11yText"
      id="react-select-7-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      className="css-1f43avz-a11yText-A11yText"
    />
    <div
      className="hods-multi-select-autocomplete__control css-x19aeh-control"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className="hods-multi-select-autocomplete__value-container hods-multi-select-autocomplete__value-container--has-value css-319lph-ValueContainer"
      >
        <div
          className="hods-multi-select-autocomplete__single-value css-qc6sy-singleValue"
        >
          bravo
        </div>
        <div
          className="hods-multi-select-autocomplete__input-container css-6j8wv5-Input"
          data-value=""
        >
          <input
            aria-autocomplete="list"
            aria-expanded={false}
            aria-haspopup={true}
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            className="hods-multi-select-autocomplete__input"
            disabled={false}
            id="react-select-7-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            role="combobox"
            spellCheck="false"
            style={
              Object {
                "background": 0,
                "border": 0,
                "color": "inherit",
                "font": "inherit",
                "gridArea": "1 / 2",
                "label": "input",
                "margin": 0,
                "minWidth": "2px",
                "opacity": 1,
                "outline": 0,
                "padding": 0,
                "width": "100%",
              }
            }
            tabIndex={0}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="hods-multi-select-autocomplete__indicators css-1hb7zxy-IndicatorsContainer"
      />
    </div>
  </div>
</div>
`;

exports[`MultiSelectAutocomplete should render an autocomplete component with an error control 1`] = `
<div
  className="hods-multi-select-autocomplete__outer-wrapper hods-multi-select-autocomplete error"
>
  <div
    className=" css-b62m3t-container"
    id="autocompleteId"
    onKeyDown={[Function]}
  >
    <span
      className="css-1f43avz-a11yText-A11yText"
      id="react-select-5-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      className="css-1f43avz-a11yText-A11yText"
    />
    <div
      className="hods-multi-select-autocomplete error__control css-x19aeh-control"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className="hods-multi-select-autocomplete error__value-container hods-multi-select-autocomplete error__value-container--has-value css-319lph-ValueContainer"
      >
        <div
          className="hods-multi-select-autocomplete error__single-value css-qc6sy-singleValue"
        >
          alpha
        </div>
        <div
          className="hods-multi-select-autocomplete error__input-container css-6j8wv5-Input"
          data-value=""
        >
          <input
            aria-autocomplete="list"
            aria-expanded={false}
            aria-haspopup={true}
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            className="hods-multi-select-autocomplete error__input"
            disabled={false}
            id="react-select-5-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            role="combobox"
            spellCheck="false"
            style={
              Object {
                "background": 0,
                "border": 0,
                "color": "inherit",
                "font": "inherit",
                "gridArea": "1 / 2",
                "label": "input",
                "margin": 0,
                "minWidth": "2px",
                "opacity": 1,
                "outline": 0,
                "padding": 0,
                "width": "100%",
              }
            }
            tabIndex={0}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="hods-multi-select-autocomplete error__indicators css-1hb7zxy-IndicatorsContainer"
      />
    </div>
  </div>
</div>
`;

exports[`MultiSelectAutocomplete should render an input with disabled flag set and the error flag set 1`] = `
<div
  className="hods-multi-select-autocomplete__outer-wrapper hods-multi-select-autocomplete error"
>
  <div
    className="hods-multi-select-autocomplete error--is-disabled css-3iigni-container"
    id="autocompleteId"
    onKeyDown={[Function]}
  >
    <span
      className="css-1f43avz-a11yText-A11yText"
      id="react-select-4-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      className="css-1f43avz-a11yText-A11yText"
    />
    <div
      className="hods-multi-select-autocomplete error__control hods-multi-select-autocomplete error__control--is-disabled css-13ibep8-control"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className="hods-multi-select-autocomplete error__value-container css-319lph-ValueContainer"
      >
        <div
          className="hods-multi-select-autocomplete error__placeholder css-14el2xx-placeholder"
          id="react-select-4-placeholder"
        >
          Select...
        </div>
        <div
          className="hods-multi-select-autocomplete error__input-container css-jzldcf-Input"
          data-value=""
        >
          <input
            aria-autocomplete="list"
            aria-describedby="react-select-4-placeholder"
            aria-expanded={false}
            aria-haspopup={true}
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            className="hods-multi-select-autocomplete error__input"
            disabled={true}
            id="react-select-4-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            role="combobox"
            spellCheck="false"
            style={
              Object {
                "background": 0,
                "border": 0,
                "color": "inherit",
                "font": "inherit",
                "gridArea": "1 / 2",
                "label": "input",
                "margin": 0,
                "minWidth": "2px",
                "opacity": 1,
                "outline": 0,
                "padding": 0,
                "width": "100%",
              }
            }
            tabIndex={0}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="hods-multi-select-autocomplete error__indicators css-1hb7zxy-IndicatorsContainer"
      />
    </div>
  </div>
</div>
`;

exports[`MultiSelectAutocomplete should render when disabled flag is set 1`] = `
<div
  className="hods-multi-select-autocomplete__outer-wrapper hods-multi-select-autocomplete"
>
  <div
    className="hods-multi-select-autocomplete--is-disabled css-3iigni-container"
    id="autocompleteId"
    onKeyDown={[Function]}
  >
    <span
      className="css-1f43avz-a11yText-A11yText"
      id="react-select-2-live-region"
    />
    <span
      aria-atomic="false"
      aria-live="polite"
      aria-relevant="additions text"
      className="css-1f43avz-a11yText-A11yText"
    />
    <div
      className="hods-multi-select-autocomplete__control hods-multi-select-autocomplete__control--is-disabled css-13ibep8-control"
      onMouseDown={[Function]}
      onTouchEnd={[Function]}
    >
      <div
        className="hods-multi-select-autocomplete__value-container hods-multi-select-autocomplete__value-container--has-value css-319lph-ValueContainer"
      >
        <div
          className="hods-multi-select-autocomplete__single-value hods-multi-select-autocomplete__single-value--is-disabled css-lr9fot-singleValue"
        >
          alpha
        </div>
        <div
          className="hods-multi-select-autocomplete__input-container css-jzldcf-Input"
          data-value=""
        >
          <input
            aria-autocomplete="list"
            aria-expanded={false}
            aria-haspopup={true}
            autoCapitalize="none"
            autoComplete="off"
            autoCorrect="off"
            className="hods-multi-select-autocomplete__input"
            disabled={true}
            id="react-select-2-input"
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            role="combobox"
            spellCheck="false"
            style={
              Object {
                "background": 0,
                "border": 0,
                "color": "inherit",
                "font": "inherit",
                "gridArea": "1 / 2",
                "label": "input",
                "margin": 0,
                "minWidth": "2px",
                "opacity": 1,
                "outline": 0,
                "padding": 0,
                "width": "100%",
              }
            }
            tabIndex={0}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="hods-multi-select-autocomplete__indicators css-1hb7zxy-IndicatorsContainer"
      />
    </div>
  </div>
</div>
`;
